contrib/vmpsd.1.3.sources/vmpsd.1
.TH OPENVMPSD 1 "Version 1.1: 19 Jan 2004"
.SH NAME
vmpsd \- Open Source Cisco VMPS Server
.SH SYNOPSIS
.B "vpmsd [-d] [-a \fIaddress\fP] [-f \fIfile\fP] [-l \fIlevel\fP] [-p \fIport\fP]"

.SH DESCRIPTION
.I vmpsd
is a GPL implementation of Cisco's VMPS server. It allows for
dynamic VLAN creation based upon MAC address, and is particularly 
helpful in smaller environments where a Cisco VMPS server is not 
available or feasible.
.PP
Since Cisco has not released documentation for this protocol, 
.I vmpsd
was based entirely on watching traffic between Cisco VMPS servers 
and clients. As such, it is not warranted to be complete or bug-free.
Despite this issue, 
.I vmpsd
is suitable for use in production environments, as it has proven to be
stable on large networks for long periods of time.

.SH OPTIONS
.TP 
.B -a " bind address" 
IP address to bind 
.I vmpsd
to. Requests from clients must come in on this address.
.TP
.B -d
Causes
.I vmpsd
to stay in the foreground instead of detaching as a daemon. Also logs
to stderr.
.TP
.B -f " config file"
Specify the configuration file that
.I vmpsd 
will use. The program will fail without this option, as currently there is no 
default set.
.TP
.B -l " log level"
Set the log level
.I vmpsd
will use. Options are:
.RS
.IP \(bu 4
.B 0x0100 
-- fatal errors only
.IP \(bu 4
.B 0x0200 
-- informational messages
.IP \(bu 4
.B 0x0400 
-- print warnings
.IP \(bu 4
.B 0x0800 
-- print debug messages; very verbose
.IP \(bu 4
.B 0x0001 
-- system
.IP \(bu 4
.B 0x0002
-- parser
.IP \(bu 4
.B 0x0004
-- vqp
.RE
.TP
.B -p " listener port"
Port
.I vmpsd
will listen on. Defaults to 1589.

.SH "FILES"
vmps.db

.SH "SEE ALSO"
.B vmpsd.db(1)

.SH AUTHORS
.PP
Program: Dori Seliskar <dori_seliskar@select-tech.si>
         David Parsley <parsley@linuxjedi.org>
.br
Manpage: Alex Kirk <alex@schnarff.com>


CHANGELOG
=====================================================================================================================
                                        Changelog for project 'FreeNAC 2.2'
=====================================================================================================================

	 * Instructions on how to import config files into the db (For 2.2 RC2) (h_ortiz)
		/doc/README.config2db.important
		/BRANCH_2.2.txt

	 * Fix syntax (h_ortiz)
		/bin/logr

	 * Add logr: monitor radius logs (h_ortiz)
		/bin/logr

	 * Hub detection (h_ortiz)
		/bin/funcs.inc
		/bin/vmpsd_external

	 * Modified version of restart_port (h_ortiz)
		/bin/restart_port.old
		/bin/funcs.inc
		/bin/restart_port

	 * v2.2.0.129/17.06.07/SB:   
  New 'Administration' tab: move Config, Users, locations, Vlans, OS to this
    new tab. Makes the interface cleaner for non-admin GUI users.
    Also add Administration->NmapSubnets to allow configuration of the subnets table.
  (source SVN 30). (boran)
		/WindowsGUI/CHANGELOG.txt
		/WindowsGUI/vmps.exe

	 * Scripts modified to use the Settings object (h_ortiz)
		/bin/funcs.inc

	 * Scripts modified to use the Settings object

	 * Small correction to is_mac_on_port function (boran)
		/bin/funcs.inc

	 * Exclude list updated (h_ortiz)
		/contrib/config2db

	 * Function remove_type (h_ortiz)
		/bin/funcs.inc

	 * vmps_components.vsd (boran)
		/doc/vmps_components.vsd

	 * Function ping_mac2 (h_ortiz)
		/bin/funcs.inc

	 * VLAN by switch location, query modified (boran)
		/bin/vmpsd_external

	 * Fixed boolean type (h_ortiz)
		/bin/Settings.php

	 * v2.2.0.127/5.06.07/SB:  
  Vlan and most other tabs: add export to excel.
  Ports: make GUI faster through optional patch lookup, add VlanAuth column
         and some color coding of columns.
  Edit-LastSeen: add Switch comment 
  Edit-LastSeen: add desired & last vlan. Add layer 3 last seen date
  Config: Add a config table for changing some server-side settings
  Users: add GuiVlanRights field: this is a CSV list of vlan indices.
    If it is set for the user starting the GUI, then only this list of vlans is visible in the edit
    tab. This is to simplifyy for 1st level support, and to limit the access to more sensitive vlans.
  Code: 
   o move all DB queries, datasources, connections to a new Datamodule, 
     for better maintenance.
   o Migrate to MyDAC5 and SDAC4 libraries (from corelabs).
  Speed: Only load the ports, giu, naclog, patch, tables when the tab in clicked on, to
         make initial loading faster. (boran)
		/WindowsGUI/vmps.exe

	 * Now read vars from the db (h_ortiz)
		/bin/Settings.php

	 * Import config files to db (h_ortiz)
		/contrib/config2db

	 * GWP detection (boran)
		/bin/vmps_lastseen

	 * GWP detection (boran)
		/bin/vmps_lastseen
		/bin/funcs.inc

	 * Add migration_2.1_to_2.2 to contrib

	 * Settings class (h_ortiz)
		/bin/Settings.php

	 * Settings class (h_ortiz)
		/bin/Settings.php

	 * Add ad_base (h_ortiz)
		/web/config.inc.template

	 * Sample for authentication in multiple domains (h_ortiz)
		/doc/webADauth.txt

	 * Support for multiple domains

	 * Increment time out value (boran)
		/bin/rad2vmps

	 * Update port regardless of check_for_expired (boran)
		/bin/vmps_lastseen

	 * v2.2.0.222/17.04.07/SB: 
  Reporting: Add 'DNS naming report' (for INO/DEMO only since its still a beta report)
  Serverlog: fix sort order. About Box: clean up, add link to website. (boran)
		/WindowsGUI/CHANGELOG.txt
		/WindowsGUI/vmps.exe

	 * Don't try to set port if we were unable to set it down (h_ortiz)
		/bin/restart_port

	 * Say whether the port was successfully restarted or not (h_ortiz)
		/bin/restart_port

	 * Do not perform Stockwerk check (boran)
		/bin/dump_ports

	 * Dump ports (boran)
		/bin/dump_ports

	 * MAC-Authentication bypass doc (h_ortiz)
		/doc/MAC-Auth-bypass.txt

	 * MAC-Authentication bypass documentation (h_ortiz)
		/doc/vmps_components.vsd
		/doc/MAC-Auth-bypass.txt

	 * Add Vmps_algorithms.vsd (boran)
		/doc/Vmps_algorithms.vsd

	 * DB name (h_ortiz)
		/WindowsGUI/demo1/vmps.xml

	 * Fixed bug in mysqlhotcopy

	 * Reverted funcs.inc (h_ortiz)
		/bin/funcs.inc

	 * Add switchname when a new switch is detected (h_ortiz)
		/bin/vmps_lastseen
		/bin/funcs.inc

	 * Removed echo (h_ortiz)
		/bin/port_scan

	 * Create different xml files for every scan (h_ortiz)
		/bin/port_scan

	 * Revision number (h_ortiz)
		/doc/webintro/index.html

	 * Don't delete file (h_ortiz)
		/bin/port_scan

	 * Add a few example vmps.xmsl. GUI changes: see change log for build 200.

	 * Add FreeNAC_training.pdf (boran)
		/doc/FreeNAC_training.pdf

	 * Added the demo DB nacdemo_db.tgz (boran)
		/contrib/nacdemo_db.tgz

	 * Rename FreeNAC_VM_howto.txt to FreeNAC_VM_8021x_howto.txt (boran)
		/doc/FreeNAC_VM_8021x_howto.txt
		/doc/FreeNAC_VM_howto.txt

	 * No debug (h_ortiz)
		/bin/vmps_lastseen

	 * Replaced vmpsd (h_ortiz)
		/bin/vmps_lastseen
		/bin/vmpsd

	 * Link to the main menu (h_ortiz)
		/web/index.php
		/web/bw.css
		/web/funcs.inc

	 * README: notes on using the demo database in the VM (boran)
		/WindowsGUI/README.WinGUI.txt

	 * Web page for the VM

	 * VM stuff (h_ortiz)
		/bin/vmpsd_external

	 * Debug off (h_ortiz)
		/bin/vmpsd_external

	 * Modifications to handle vms (h_ortiz)
		/bin/funcs.inc
		/bin/vmpsd_external

	 * vmps.exe 119 (boran)
		/WindowsGUI/vmps.exe

	 * vmps.xml: default DB is not opennac, not inventory (boran)
		/WindowsGUI/vmps.xml

	 * Added hostname when warning about expired (boran)
		/bin/vmps_lastseen
		/bin/port_scan

	 * Check for unknown status (h_ortiz)
		/bin/vmpsd_external

	 * is_field_active function (h_ortiz)
		/bin/funcs.inc

	 * Remove vars that control the disabling of expired systems (h_ortiz)
		/bin/vmpsd_external

	 * virtual machines

	 * SB: remove config_static, mv mytop to usr/local/bin (boran)
		/contrib/usr/local/bin/mytop
		/contrib/README.config_static
		/contrib/mytop

	 * SB: add rdiff.readme (boran)
		/contrib/rdiff.readme

contrib/vmpsd.1.3.sources/vmpsd.c
#include "config.h"

#include <stdlib.h>
#include <signal.h>

#if HAVE_UNISTD_H
#include <unistd.h>
#endif

#include "vqp.h"
#include "log.h"
#include "external.h"

struct in_addr	bind_address; 
unsigned int	port_number = 1589;
char    	db_fname[256];

int parse_options(int argc, char **argv)
{
	char	opt;
	char	*options = "a:de:f:l:p:";

	opterr = 0;
	
	opt = getopt(argc, argv, options);
	while ( opt > 0 ) {

		switch (opt) {

			case 'a':
				if ( optarg == NULL ) return 0;
				if ( !inet_aton(optarg, &bind_address) ) return 0;	
				break;

			case 'd':
				debug = 1;
				break;

			case 'e':
				strncpy(external_prog, optarg, 255);
				external_prog[255] = '\0';
				external_logic = 1;
				break;
				
			case 'f':
				strncpy(db_fname, optarg, 255);
				db_fname[255] = '\0';
				break;
				
			case 'l':
				if ( sscanf(optarg,"%x",&log_level) != 1) return 0;
				break;

			case 'p':
				if ( sscanf(optarg,"%d",&port_number) != 1) return 0;
				break;

			default:
				return 0;
				break;
		}
		opt = getopt(argc, argv, options);
	}

	return 1;
}

void usage()
{
	printf("\n");
	printf("Options:\n");
	printf("\n");
	printf("\t-a ip      address to bind to (any)\n");
	printf("\t-d         do not detach, log to stderr also\n");
	printf("\t-e path    use external program for mac to vlan assignment\n");
	printf("\t           when/if used with -f, -f is disregarded\n");
	printf("\t-f file    read VMPS database from file (/etc/vmps.db)\n");
	printf("\t-l level   set logging level:\n");
	printf("\t                 0x0100 - fatal,\n");
	printf("\t                 0x0200 - info,\n");
	printf("\t                 0x0400 - warning,\n");
	printf("\t                 0x0800 - debug,\n");
	printf("\t                 0x0001 - system,\n");
	printf("\t                 0x0002 - parser,\n");
	printf("\t                 0x0004 - vqp\n");
	printf("\t-p port    port to listen on (1589)\n");
	printf("\n");
}

RETSIGTYPE handle_sighup() {

	if ( external_logic ) return;
	vmps_log(PARSER|INFO, "RECEIVED SIGHUP. Re-reading config file");
	drop_data();
	parse_db_file(db_fname);
}

int main(int argc, char **argv) 
{
	int			sock;
	struct sockaddr_in	serv_addr;

	VQP_REQUEST		r;

	bind_address.s_addr = INADDR_ANY;
	strncpy(db_fname,SYSCONFDIR,240);
	strcat(db_fname,"/vlan.db");

	if ( !parse_options(argc,argv) ) {
		usage();
		exit(1);
	}

	if ( !debug ) daemon_start(1); 

	if ( external_logic ) {
		spawn_external();
	} else {
		parse_db_file(db_fname);
	}
	
	if ( (sock = socket(AF_INET, SOCK_DGRAM, 0)) < 0 ) {
		vmps_log(FATAL|SYSTEM, "Cannot create a socket.");
		exit(1);
	}

	bzero( (char *) &serv_addr, sizeof(serv_addr) );
	serv_addr.sin_family	 = AF_INET;
	serv_addr.sin_addr.s_addr = bind_address.s_addr;
	serv_addr.sin_port	 = htons(port_number);

	if ( bind(sock, (struct sockaddr *) &serv_addr, sizeof(serv_addr)) < 0 ) {
		vmps_log(FATAL|SYSTEM, "Cannot bind the socket.");
		exit(1);
	}

#ifdef HAVE_SIGACTION
	{
		struct sigaction action;
		action.sa_sigaction = handle_sighup;
		sigemptyset(&action.sa_mask);
		action.sa_flags = SA_SIGINFO;
		sigaction(SIGHUP, &action, NULL);
	}
#else
		signal(SIGHUP, handle_sighup);
#endif

	vmps_log(SYSTEM|INFO, "VMPSD STARTED. Waiting for requests");

	while (1) {
		
		if ( !get_request(sock, &r) ) {
			if ( (log_level & 0xFF00) >= DEBUG ) print_request(&r);
			
			if ( external_logic ) 	{ do_request_external(sock, &r); }
			else			{ do_request(sock, &r); }

		}
	}
}


